xds/internal/balancer/outlierdetection/balancer.go

```diff
@@ -105,3 +132,760 @@ func (bb) ParseConfig(s json.RawMessage) (serviceconfig.LoadBalancingConfig, err
+		return
+	}
+
+	// 2. Calculate the mean and standard deviation of the fractions of
+	// successful requests among addresses with total request volume of at least
+	// success_rate_ejection.request_volume.
+	mean, stddev := b.meanAndStdDevOfSuccessesAtLeastRequestVolume()
+
+	// 3. For each address:
+	for addr, obj := range b.addrs {
+		ccb := obj.callCounter.inactiveBucket
+		sre := b.cfg.SuccessRateEjection
+		// i. If the percentage of ejected addresses is greater than
+		// max_ejection_percent, stop.
+		if float64(b.numAddrsEjected)/float64(len(b.addrs))*100 > float64(b.cfg.MaxEjectionPercent) {
+			return
+		}
+
+		// ii. If the address's total request volume is less than
+		// success_rate_ejection.request_volume, continue to the next address.
+		if ccb.requestVolume < int64(sre.RequestVolume) {
+			continue
+		}
+
+		//  iii. If the address's success rate is less than (mean - stdev *
+		//  (success_rate_ejection.stdev_factor / 1000))
+		successRate := float64(ccb.numSuccesses) / float64(ccb.requestVolume)
+		if successRate < (mean - stddev*(float64(sre.StdevFactor)/1000)) {
+			// then choose a random integer in [0, 100). If that number is less
+			// than success_rate_ejection.enforcement_percentage, eject that
+			// address.
+			if uint32(grpcrand.Int31n(100)) < sre.EnforcementPercentage {
+				b.ejectAddress(addr)
+			}
+		}
+	}
+}
+
+func (b *outlierDetectionBalancer) failurePercentageAlgorithm() {
+	// 1. If the number of addresses is less than
+	// failure_percentage_ejection.minimum_hosts, stop.
+	if uint32(len(b.addrs)) < b.cfg.FailurePercentageEjection.MinimumHosts {
+		return
+	}
+
+	// 2. For each address:
+	for addr, obj := range b.addrs {
+		ccb := obj.callCounter.inactiveBucket
+		fpe := b.cfg.FailurePercentageEjection
+		// i. If the percentage of ejected addresses is greater than
+		// max_ejection_percent, stop.
+		if float64(b.numAddrsEjected)/float64(len(b.addrs))*100 > float64(b.cfg.MaxEjectionPercent) {
+			return
+		}
+		// ii. If the address's total request volume is less than
+		// failure_percentage_ejection.request_volume, continue to the next
+		// address.
+		if uint32(ccb.requestVolume) < fpe.RequestVolume {
+			continue
+		}
+		//  2c. If the address's failure percentage is greater than
+		//  failure_percentage_ejection.threshold
+		failurePercentage := (float64(ccb.numFailures) / float64(ccb.requestVolume)) * 100
+		if failurePercentage > float64(b.cfg.FailurePercentageEjection.Threshold) {
+			// then choose a random integer in [0, 100). If that number is less
+			// than failiure_percentage_ejection.enforcement_percentage, eject
+			// that address.
+			if uint32(grpcrand.Int31n(100)) < b.cfg.FailurePercentageEjection.EnforcementPercentage {
+				b.ejectAddress(addr)
+			}
+		}
+	}
+}
+
+func (b *outlierDetectionBalancer) ejectAddress(addr string) {
+	obj, ok := b.addrs[addr]
+	if !ok { // Shouldn't happen
```

Comment:
Can we pass `b.addrs[addr]` here (and `unejectAddress`) instead to guarantee this won't happen?